#Introduction-------------------------------------------------------------------

#This script aims to analyse Rattus norvegicus RNA seq data generated by
#Filippi's lab. This will be achieved by performing differentially expressed
#gene analysis using DESeq2.

#RStudio version 2024.04.2 Build 764 (Chocolate Cosmos).
#R version 4.4.1 (Race for Your Life).
#Confirmed to work on Windows 11.

#Developed by Azan Ahmed 03/05/2024.

#Prepare for analysis-----------------------------------------------------------

library(tximport)
library(DESeq2)
library(org.Rn.eg.db)
library(ggplot2)

# Import files using tximport---------------------------------------------------

setwd("path/to/your/data")

quant_files <- c(
  "GFP 1" = "1-GFP_Ctrl_RNA_S1_L001_R1_001/quant_1.sf",
  "GFP 2" = "2-GFP_Ctrl_RNA_S2_L001_R1_001/quant_2.sf",
  "GFP 3" = "3-GFP_Ctrl_RNA_S3_L001_R1_001/quant_3.sf",
  "GFP + Insulin 1" =  "4-GFP_Ins_RNA_S4_L001_R1_001/quant_4.sf",
  "GFP + Insulin 2" =  "5-GFP_Ins_RNA_S5_L001_R1_001/quant_5.sf",
  "GFP + Insulin 3" =  "6-GFP_Ins_RNA_S6_L001_R1_001/quant_6.sf",
  "Drp-1 CA 1" =  "7-CrP_Ctrl_RNA_S7_L001_R1_001/quant_7.sf",
  "Drp-1 CA 2" =  "8-CrP_Ctrl_RNA_S8_L001_R1_001/quant_8.sf",
  "Drp-1 CA 3" =  "9-CrP_Ctrl_RNA_S9_L001_R1_001/quant_9.sf",
  "Drp-1 CA + Insulin 1" =  "10-CrP_Ins_RNA_S10_L001_R1_001/quant_10.sf",
  "Drp-1 CA + Insulin 2" =  "11-CrP_Ins_RNA_S11_L001_R1_001/quant_11.sf",
  "Drp-1 CA + Insulin 3" =  "12-CrP_Ins_RNA_S12_L001_R1_001/quant_12.sf"
)

gtf_file <- rtracklayer::import( "GCF_036323735.1_GRCr8_genomic.gtf.gz")

tx2gene_file <- unique(
  data.frame(gtf_file[gtf_file$type == "exon"])[,c("transcript_id", "db_xref")])

txi <- tximport(quant_files,
                type = "salmon",
                tx2gene = tx2gene_file)

#Differential expression analysis-----------------------------------------------
sample_table <- data.frame(
  sample = c(
    "GFP 1",
    "GFP 2",
    "GFP 3",
    "GFP + Insulin 1",
    "GFP + Insulin 2",
    "GFP + Insulin 3",
    "Drp-1 CA 1",
    "Drp-1 CA 2",
    "Drp-1 CA 3",
    "Drp-1 CA + Insulin 1",
    "Drp-1 CA + Insulin 2",
    "Drp-1 CA + Insulin 3"
  ),
  condition = c(
    rep("GFP", 3),
    rep("GFP + Insulin", 3),
    rep("Drp-1 CA", 3),
    rep("Drp-1 CA + Insulin", 3)
  ),
  Treatment = c(
    rep("Control", 3),
    rep("Insulin", 3),
    rep("Control", 3),
    rep("Insulin", 3)
  ),
  Genotype = c(rep("GFP", 6), rep("Drp-1 CA", 6)
  ))

dds <- DESeqDataSetFromTximport(txi, 
                                colData = sample_table, 
                                design = ~ condition)
dds <- DESeq(dds)

#Obtain list of genes detected in RNA-seq for Overrepresentation analysis.
gene_universe <- rownames(dds)
gene_universe <- sub("GeneID:", "", gene_universe)
saveRDS(gene_universe, file = "gene_universe.rds")

#For PCA and Heat map
rld <- rlog(dds, blind = FALSE)

#PCA----------------------------------------------------------------------------
#Ensure for rld blind was set as TRUE.
pca_data <- plotPCA(rld,
                    intgroup = c("Treatment", "Genotype"),
                    returnData = TRUE)

#Round PC variance
pc_percent <- round(100 * attr(pca_data, "percentVar"))

#remove_grid function
remove_grid <- 
  function(x = TRUE, y = TRUE) {
  p <- ggplot2::theme(panel.grid.minor = ggplot2::element_blank())
  if (x) {
    p <- p +
      ggplot2::theme(panel.grid.major.x = ggplot2::element_blank())
  }
  if (y) {
    p <- p +
      ggplot2::theme(panel.grid.major.y = ggplot2::element_blank())
  }
    
  p
}

pca_palette <- c("Control" = "#009E73", "Insulin" = "#56B4E9")

#Plot PCA graph
ggplot(pca_data, aes(x = PC1, y = PC2, fill = Treatment, shape = Genotype)) +
  geom_point(size = 3, stroke = 0.5, alpha = 0.7) +
  theme_light(base_size = 15) +
  theme(panel.border = element_blank(),
        axis.line = element_line(colour = "black"),
        axis.ticks = element_line(color = "black")) +
  xlab(paste0("PC1: ", pc_percent[1], "% variance")) +
  ylab(paste0("PC2: ", pc_percent[2], "% variance")) +
  scale_y_continuous(limits = c(-10, 10), expand = expansion(mult = c(0, 0))) +
  scale_x_continuous(limits = c(-15, 15), expand = expansion(mult = c(0, 0))) +
  coord_fixed() +
  scale_fill_manual(values =pca_palette) +
  scale_shape_manual(values = c("GFP" = 21, "Drp-1 CA" = 24)) +
  labs(fill = "Treatment", shape = "Viral modification") +
  guides(fill = guide_legend(override.aes = list(color = pca_palette)),
         shape = guide_legend(override.aes = list(color = "black"))) +
  remove_grid()

#Gene clustering----------------------------------------------------------------
#Ensure for rld blind was set to FALSE.
library(ComplexHeatmap)
library(genefilter)
library(RColorBrewer)
  
top_variable_genes <- head(order(rowVars(assay(rld)), decreasing = TRUE), 50)
mat <- assay(rld)[ top_variable_genes, ]
rownames(mat) <- sub("GeneID:", "", rownames(mat))
gene_ids <- rownames(mat)
  
gene_symbols <- mapIds(
  org.Rn.eg.db,
  keys = gene_ids,
  column = "SYMBOL",
  keytype = "ENTREZID",
  multiVals = "first"
)

rownames(mat) <- gene_symbols[rownames(mat)]
anno <- as.data.frame(colData(rld)[, c("Treatment", "Genotype")])
colnames(anno) <- c("Treatment", "Viral Modification")
annotation_colors <- list(
  Treatment = c("Control" = "#009E73", "Insulin" = "#56B4E9"),
  "Viral Modification" = c("GFP" = "#CC79A7", "Drp-1 CA" = "#0072B2")
)


#Not scaled Heat map.
pheatmap(
  mat,
  annotation_col = anno,
  name = " ",
  color = colorRampPalette(c("#F9F9F9", "#79ABE2", "#00366C"))(50),
  annotation_colors = annotation_colors
)

#Z-score scaled.
mat_scaled <- t(scale(t(mat)))

#Z-score scaled heat map.
pheatmap(
  mat_scaled,
  annotation_col = anno,
  name = "Z-Score",
  color = colorRampPalette(c("#4A6FE3", "#E2E2E2", "#D33F6A"))(50),
  annotation_colors = annotation_colors,
  cluster_rows = FALSE
)

#Export results for Control vs Insulin------------------------------------------

library(apeglm)#For lfcShrink command

sample_table$condition <- factor(
  sample_table$condition,
  levels = c("GFP", "GFP + Insulin", "Drp-1 CA", "Drp-1 CA + Insulin")
)

dds$condition <- relevel(sample_table$condition, ref = "GFP")
dds <- nbinomWaldTest(dds)
resultsNames(dds)
dds_results_1 <- lfcShrink(
  dds,
  coef = "condition_GFP...Insulin_vs_GFP",
  type = "apeglm")

dds_results_1_df <- as.data.frame(dds_results_1)
dds_results_1_df <- na.omit(dds_results_1_df)
rownames(dds_results_1_df) <- sub("GeneID:", "", rownames(dds_results_1_df))
ncbi_gene_id_1 <-  rownames(dds_results_1_df)
gene_symbols_1 <- mapIds(org.Rn.eg.db,
                         keys = ncbi_gene_id_1,
                         column = "SYMBOL",
                         keytype = "ENTREZID",
                         multiVals = "first")
dds_results_1_df$GeneSymbol <- gene_symbols_1
write.csv(dds_results_1_df, file="Control_vs_Insulin.csv")

#Export results for Control vs Drp-1 CA-----------------------------------------
dds$condition <- relevel(sample_table$condition, ref = "GFP")
dds <- nbinomWaldTest(dds)
resultsNames(dds)
dds_results_2 <- lfcShrink(
  dds,
  coef = "condition_Drp.1.CA_vs_GFP",
  type = "apeglm")

dds_results_2_df <- as.data.frame(dds_results_2)
dds_results_2_df <- na.omit(dds_results_2_df)
rownames(dds_results_2_df) <- sub("GeneID:", "", rownames(dds_results_2_df))
ncbi_gene_id_2 <-  rownames(dds_results_2_df)
gene_symbols_2 <- mapIds(org.Rn.eg.db,
                         keys = ncbi_gene_id_2,
                         column = "SYMBOL",
                         keytype = "ENTREZID",
                         multiVals = "first")
dds_results_2_df$GeneSymbol <- gene_symbols_2
write.csv(dds_results_2_df, file="Control_vs_Drp-1_CA.csv")

#Export results for Insulin vs Drp-1 CA + Insulin-------------------------------
dds$condition <- relevel(sample_table$condition, ref = "GFP + Insulin")
dds <- nbinomWaldTest(dds)
resultsNames(dds)
dds_results_3 <- lfcShrink(
  dds,
  coef = "condition_Drp.1.CA...Insulin_vs_GFP...Insulin",
  type = "apeglm")

dds_results_3_df <- as.data.frame(dds_results_3)
dds_results_3_df <- na.omit(dds_results_3_df)
rownames(dds_results_3_df) <- sub("GeneID:", "", rownames(dds_results_3_df))
ncbi_gene_id_3 <-  rownames(dds_results_3_df)
gene_symbols_3 <- mapIds(org.Rn.eg.db,
                         keys = ncbi_gene_id_3,
                         column = "SYMBOL",
                         keytype = "ENTREZID",
                         multiVals = "first")
dds_results_3_df$GeneSymbol <- gene_symbols_3
write.csv(dds_results_3_df, file="Insulin_vs_Drp-1_CA_Insulin.csv")

#Export results for Drp-1 CA vs Drp-1 CA + Insulin------------------------------
dds$condition <- relevel(sample_table$condition, ref = "Drp-1 CA")
dds <- nbinomWaldTest(dds)
resultsNames(dds)
dds_results_4_df <- lfcShrink(
  dds,
  coef = "condition_Drp.1.CA...Insulin_vs_Drp.1.CA",
  type = "apeglm")

dds_results_4_df <- na.omit(dds_results_4_df)
rownames(dds_results_4_df) <- sub("GeneID:", "", rownames(dds_results_4_df))
ncbi_gene_id_4 <-  rownames(dds_results_4_df)
gene_symbols_4 <- mapIds(org.Rn.eg.db,
                         keys = ncbi_gene_id_4,
                         column = "SYMBOL",
                         keytype = "ENTREZID",
                         multiVals = "first")
dds_results_4_df$GeneSymbol <- gene_symbols_4
write.csv(dds_results_4_df, file="Drp-1_vs_Drp-1_CA_Insulin.csv")

#MA Plots-----------------------------------------------------------------------

#Change for each comparison
plotMA(dds_results_1, ylim = c(-5, 5))

#Single Gene Plots--------------------------------------------------------------
library(dplyr)

remove_grid_x <- function() {
  remove_grid(x = TRUE, y = FALSE)
}

gene_of_interest <- plotCounts(dds,
                            "GeneID:767613",
                            returnData = TRUE,
                            intgroup = "condition")

gene_of_interest$Genotype <- ifelse(
  grepl("GFP", gene_of_interest$condition),
  "GFP",
  "Drp-1 CA")

gene_of_interest$Genotype <- factor(
  gene_of_interest$Genotype,
  levels = c("GFP", "Drp-1 CA"))

gene_of_interest$Treatment <- ifelse(
  grepl("Insulin", gene_of_interest$condition),
  "Insulin",
  "Control")

#If you only want to plot control samples.
gene_of_interest_control <- gene_of_interest %>%
  filter(Treatment == "Control")

bar_fill_colors <- c("Control" = "#009E73", "Insulin" = "#56B4E9")
bar_outline_colors <- c("Control" = "#005841", "Insulin" = "#2A6E98")

summary_data <- gene_of_interest_control %>%
  group_by(Genotype, Treatment) %>%
  summarise(
    mean = mean(count),
    SEM = sd(count) / sqrt(n())
  )

ggplot(summary_data, aes(x = Genotype, y = Mean, fill = Treatment)) +
  geom_bar(stat = "identity", width = 0.8, color = "black", 
           position = position_dodge(width = 0.8)) +
  geom_errorbar(
    aes(ymin = mean - SEM, ymax = mean + SEM),
    width = 0.2,
    position = position_dodge(width = 0.8)
  ) +
  geom_point(
    data = gene_of_interest_control,
    aes(x = Genotype, y = count, color = Treatment),
    position = position_jitterdodge(jitter.width = 0.2, dodge.width = 0.8),
    size = 2,
    shape = 4,
    stroke = 1
  ) +
  labs(x = "Viral modification", y = "Normalized Counts") +
  theme_light(base_size = 10) +
  theme(
    axis.line = element_line(colour = "black"),
    axis.ticks.x = element_blank(),
    panel.border = element_blank(),
    axis.ticks.y = element_line(colour = "black"),
    legend.position = "none"
  ) +
  scale_y_continuous(expand = expansion(mult = c(0, 0))) +
  scale_fill_manual(values = bar_fill_colors) +
  scale_color_manual(values = bar_outline_colors) +
  removeGrid() +
  coord_cartesian(ylim = c(0, 15000))


